HW2/P5/physics.pyx

```diff
@@ -113,18 +160,36 @@ cpdef update(FLOAT[:, ::1] XY,
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
         # scheduling.
-        for i in range(count):
-            sub_update(XY, V, R, i, count, Grid, grid_spacing)
+        # for i in range(count):
+        for i in prange(count, schedule='static', chunksize=count/4, num_threads=4):
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, grid_size, locks)
 
         # update positions
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
         #    scheduling).
         # SUBPROBLEM 2: update the grid values.
-        for i in range(count):
+        
+        # Reinitialize the grid with no balls
+        Grid[:,:] = -1
+
+        # for i in range(count):
+        for i in prange(count, schedule='static', chunksize=count/4, num_threads=4):
+            
             for dim in range(2):
                 XY[i, dim] += V[i, dim] * t
-
+                
+                # After updating, check to see if any positions are greater than [0,1]
+                if XY[i, dim] < 0:
+                    XY[i, dim] = 0
+                if XY[i,0] > 1:
+                    XY[i, dim] = 1
+
+            # Update the ball positions in the Grid
+            grid_pos_x = <int> (XY[i, 0]/grid_spacing)
+            grid_pos_y = <int> (XY[i, 1]/grid_spacing)
+            Grid[grid_pos_x , grid_pos_y] = i
+            
 
```

Comment:
I appreciated the high number of comments within the code. It made it much easier to follow the code